File: src/prng.cpp
#include "minml/prng.h"

#include "threefry.h"

namespace minml {

PRNGKey PRNGKey::from_seed(uint32_t seed) {
  // Hash the seed once so seed=0 doesn't produce a degenerate key.
  uint32_t a, b;
  threefry_2x32(seed, 0xCAFEBABE, 0, 0, a, b);
  return PRNGKey(a, b);
}

std::vector<PRNGKey> PRNGKey::split(size_t n) const {
  std::vector<PRNGKey> out;
  out.reserve(n);
  for (size_t i = 0; i < n; ++i) {
    uint32_t a, b;
    threefry_2x32(k0_, k1_, static_cast<uint32_t>(i), 0xC0FFEE, a, b);
    out.emplace_back(a, b);
  }
  return out;
}

}  // namespace minml

File: src/prng.ts
import { threefry_2x32 } from "./threefry.js";

export class PRNGKey {
  readonly k0: number;
  readonly k1: number;

  constructor(k0: number, k1: number) {
    this.k0 = k0 >>> 0;
    this.k1 = k1 >>> 0;
  }

  // Hash the seed once so seed=0 doesn't produce a degenerate key.
  static from_seed(seed: number): PRNGKey {
    const [a, b] = threefry_2x32(seed >>> 0, 0xcafebabe, 0, 0);
    return new PRNGKey(a, b);
  }

  // n derived keys; same parent always yields the same children.
  split(n: number): PRNGKey[] {
    const out: PRNGKey[] = [];
    for (let i = 0; i < n; i++) {
      const [a, b] = threefry_2x32(this.k0, this.k1, i >>> 0, 0xc0ffee);
      out.push(new PRNGKey(a, b));
    }
    return out;
  }

  equals(other: PRNGKey): boolean {
    return this.k0 === other.k0 && this.k1 === other.k1;
  }
}
